# Исследование надежности заемщиков.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления.</a></span><ul class="toc-item"><li><span><a href="#Отрицательные-значения-в-столбце-days_employed" data-toc-modified-id="Отрицательные-значения-в-столбце-days_employed-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Отрицательные значения в столбце <code>days_employed</code></a></span></li><li><span><a href="#Положительные-значения-в-столбце-days_employed" data-toc-modified-id="Положительные-значения-в-столбце-days_employed-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Положительные значения в столбце <code>days_employed</code></a></span></li><li><span><a href="#Сравнение-положительных-и-отрицательных-значений" data-toc-modified-id="Сравнение-положительных-и-отрицательных-значений-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Сравнение положительных и отрицательных значений</a></span></li></ul></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span><ul class="toc-item"><li><span><a href="#Явные-дубликаты" data-toc-modified-id="Явные-дубликаты-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Явные дубликаты</a></span></li><li><span><a href="#Неявные-дубликаты" data-toc-modified-id="Неявные-дубликаты-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Неявные дубликаты</a></span></li><li><span><a href="#Методы-обработки-дубликатов" data-toc-modified-id="Методы-обработки-дубликатов-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Методы обработки дубликатов</a></span></li></ul></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:" data-toc-modified-id="Вопрос-1:-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Вопрос 1:</a></span><ul class="toc-item"><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-9.1.1"><span class="toc-item-num">9.1.1&nbsp;&nbsp;</span>Вывод 1:</a></span></li></ul></li><li><span><a href="#Вопрос-2:" data-toc-modified-id="Вопрос-2:-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Вопрос 2:</a></span><ul class="toc-item"><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-9.2.1"><span class="toc-item-num">9.2.1&nbsp;&nbsp;</span>Вывод 2:</a></span></li></ul></li><li><span><a href="#Вопрос-3:" data-toc-modified-id="Вопрос-3:-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Вопрос 3:</a></span><ul class="toc-item"><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-9.3.1"><span class="toc-item-num">9.3.1&nbsp;&nbsp;</span>Вывод 3:</a></span></li></ul></li><li><span><a href="#Вопрос-4:" data-toc-modified-id="Вопрос-4:-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>Вопрос 4:</a></span><ul class="toc-item"><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-9.4.1"><span class="toc-item-num">9.4.1&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Шаг 1. Обзор данных

   Нами получены данные от заказчика - кредитного отдела банка. Данные предоставлены в виде `.csv` файла, содержащего статистическую информацию о платёжеспособности клиентах банка. Необходимо разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

Рассмотрим полученные данные.

In [1]:
import pandas as pd
import warnings

In [2]:
data = pd.read_csv ('data.csv')
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [3]:
data.info()
data.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

В представленной информации о датасете заметны два столбца с явными пропусками значений:
1. Столбец `days_employed` - столбец, отображающий количество дней трудового стажа.
2. Столбец `total_income` - столбец, отображающий общий доход клиента.

Кроме того, вызывают сомнения отрицательные значения в столбце `days_employed`.

**Приступим к предобработке данных**

## Шаг 2.1 Заполнение пропусков

В первую очередь займемся явными попусками значений в датасете. 

In [4]:
print('Количество пропущенных значений в столбце \'total_income\'-', data['total_income'].isna().sum())
print('Количество пропущенных значений в столбце \'days_employed\'-', data['days_employed'].isna().sum())
print()
total_income_share =  data['total_income'].isna().sum()/len(data)
days_employed_share = data['days_employed'].isna().sum()/len(data)
print(f'Процент пропущенных значений в столбце \'total_income\'- {total_income_share:.2%}')
print(f'Процент пропущенных значений в столбце \'days_employed\'- {days_employed_share:.2%}')

Количество пропущенных значений в столбце 'total_income'- 2174
Количество пропущенных значений в столбце 'days_employed'- 2174

Процент пропущенных значений в столбце 'total_income'- 10.10%
Процент пропущенных значений в столбце 'days_employed'- 10.10%


Количество пропущенных значений и их процент от общего объема данных в обоих столбцах совпадает. Можно предположить, что они находятся в одних и тех же строках. Убедимся в этом, выведя строки, где пропущены значения хотя бы в одном из столбцов.

In [5]:
data[(data['total_income'].isna()== True) | (data['days_employed'].isna() == True)].head()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [6]:
data[(data['total_income'].isna()== True) | (data['days_employed'].isna() == True)].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2174 entries, 12 to 21510
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          2174 non-null   int64  
 1   days_employed     0 non-null      float64
 2   dob_years         2174 non-null   int64  
 3   education         2174 non-null   object 
 4   education_id      2174 non-null   int64  
 5   family_status     2174 non-null   object 
 6   family_status_id  2174 non-null   int64  
 7   gender            2174 non-null   object 
 8   income_type       2174 non-null   object 
 9   debt              2174 non-null   int64  
 10  total_income      0 non-null      float64
 11  purpose           2174 non-null   object 
dtypes: float64(2), int64(5), object(5)
memory usage: 220.8+ KB


Первоначальное предположение подтвердилось: значения в обоих столбцах пропущены в одних тех же строках.
Исходя из полученых данных, можно выдвинуть гипотезу о возникновении пропусков:
Возможно в программе формировавшей датасет эти два столбца были связаны между собой функцией, которая рассчитывала значения одного из столбцов, исходя из значений другого. Тогда наиболее вероятно, что значения в одном из столбцов не были заполнены сотрудником банка (*человеческий фактор*), в результате чего программа не смогла рассчитать значенеия для другого стлбца (*технологическая ошибка*).

**Заполним пропущенные значения в столбце `total_income` медианным значением.**

Так как значения в этом столбце являются количественными, то пропуски в них лучше заполнять характерными значениями, характеризующими состояние выборки. Чтобы примерно оценить типичные значения выборки, годятся среднее арифметическое или медиана. Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства. Для большой выборки предпочтительно использовать медиану.

In [7]:
data['total_income']=data['total_income'].fillna(data['total_income'].median())
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Пропуски в столбце `total_income` устранены.

## Шаг 2.2 Проверка данных на аномалии и исправления.

В столбце `days_employed`, характеризующем общий трудовой стаж в днях, как было замечено ранее, присутсвуют отрицательные значения. 

Оценим расброс в значениях между отрицательными и положительными.

In [8]:
print('Минимальное значение количества дней трудового стажа -', data['days_employed'].min())
print('Максимальное значение количества дней трудового стажа -', data['days_employed'].max())

Минимальное значение количества дней трудового стажа - -18388.949900568383
Максимальное значение количества дней трудового стажа - 401755.40047533


Так как количественный показатель дней трудового стажа не может быть меньше 0, рассмотрим показатели отдельно: "отрицательные" и "положительные". 

### Отрицательные значения в столбце `days_employed`

In [9]:
negative_median = data[data['days_employed']<0]['days_employed'].median()
negative_min = data[data['days_employed']<0]['days_employed'].min()
negative_max = data[data['days_employed']<0]['days_employed'].max()
print('Медианное значение - ', round(negative_median, 2))
print('Минимальное значение - ', round(negative_min, 2))
print('Максимальное значение - ', round(negative_max, 2))

Медианное значение -  -1630.02
Минимальное значение -  -18388.95
Максимальное значение -  -24.14


Виден большой разброс значений между минимальным и максимальным значениями.

Для удобства работы с данными будем использовать для расчетов абсолютные значения. 
Попоробуем рассчитать количество лет трудового стажа по полученным значениям дней.

In [10]:
print('Медианное значение лет - ', round(abs(negative_median)/365, 2))
print('Минимальное значение лет - ', round(abs(negative_min)/365, 2))
print('Максимальное значение лет - ', round(abs(negative_max)/365, 2))

Медианное значение лет -  4.47
Минимальное значение лет -  50.38
Максимальное значение лет -  0.07


В данной выборке значений года трудого стажа похожи на реальные значения.

### Положительные значения в столбце `days_employed`

In [11]:
positive_median = data[data['days_employed']>=0]['days_employed'].median()
positive_min = data[data['days_employed']>=0]['days_employed'].min()
positive_max = data[data['days_employed']>=0]['days_employed'].max()
print('Медианное значение - ', round(positive_median, 2))
print('Минимальное значение - ', round(positive_min, 2))
print('Максимальное значение - ',round(positive_max, 2))

Медианное значение -  365213.31
Минимальное значение -  328728.72
Максимальное значение -  401755.4


Значения более однородны. Разница между минимальным и максимальным значением не столь очевидна, значение медианы приближено к среднему значению.

Рассчитаем количество лет трудового стажа по полученным значениям дней.

In [12]:
print('Медианное значение лет - ', round(abs(positive_median)/365, 2))
print('Минимальное значение лет - ', round(abs(positive_min)/365, 2))
print('Максимальное значение лет - ', round(abs(positive_max)/365, 2))

Медианное значение лет -  1000.58
Минимальное значение лет -  900.63
Максимальное значение лет -  1100.7


Значения в данной выборке при пересчете на количество лет трудового стажа выглядят неправдоподобно большими.

### Сравнение положительных и отрицательных значений

Рассмотрим обе категории в сравнении для выявления закономерностей.

*Посчитаем количество и долю каждой категории значений от общего количества `non-null` значений в столбце.*

In [13]:
print('Количество отрицательных значений -', data[data['days_employed']<0]['days_employed'].count())
negative_share = data[data['days_employed']<0]['days_employed'].count()/data['days_employed'].value_counts().count()
print(f'Процент отрицательных значений - {negative_share:.2%}')
print ()
print('Количество положительных значений -', data[data['days_employed']>=0]['days_employed'].count())
positive_share = data[data['days_employed']>=0]['days_employed'].count()/data['days_employed'].value_counts().count()
print(f'Процент положительных значений - {positive_share:.2%}')

Количество отрицательных значений - 15906
Процент отрицательных значений - 82.20%

Количество положительных значений - 3445
Процент положительных значений - 17.80%


Положительных значений существенно меньше отрицательных.

*Рассмотрим количество уникальных значений по столбцу `income_type` для каждой категории*

In [14]:
'Типы дохода для отрицательных значений -', data[data['days_employed']<0]['income_type'].value_counts()

('Типы дохода для отрицательных значений -',
 сотрудник          10014
 компаньон           4577
 госслужащий         1312
 студент                1
 предприниматель        1
 в декрете              1
 Name: income_type, dtype: int64)

In [15]:
'Типы дохода для положительных значений -',data[data['days_employed']>=0]['income_type'].value_counts()

('Типы дохода для положительных значений -',
 пенсионер      3443
 безработный       2
 Name: income_type, dtype: int64)

Категория отрицательных значений стлобца `days_employed` представлена шестью типами дохода: `сотрудник`, `компаньон`, `госслужащий`, `студент`, `предприниматель`, `в декрете`.

Категория положительных значений стлобца `days_employed` представлена двумя типами дохода: `пенсионер`, `безработный`. Из них 3443 пенсионера и 2 безработных.

Исходя из проведенного выше анализа, можно сделать вывод о том, что отрицательные значения в стлобце `days_employed` отображают количество дней трудового стажа трудоустроенных клиентов похожих на реальные, но записанные в виде отрицательного значения.

В тоже время, положительные значения в стлобце `days_employed`, которые значительно превышают продолжительность жизни человека, представлены всего двумя категориями нетрудоустроенных клиентов и представляют около **18%** от общего объема данных.

Поскольку на данный момент нет возможности уточнить принцип заполнения стлобца `days_employed` и появление таких неправдоподобно больших значений, необходимо принять решение о возможности использования этих значений для дальнейшего анализа данных.

Учитывая, что дальнейший анализ не предусматривает обращение к значениям данного столбца, примем решение удалить вызывающие недоверие значения в столбце `days_employed` для дальнейшего заполнения их медианными значениями при заполнении пропущенных значений.

Удалим положительные значения из столбца `days_employed` и заменим пропуски в столбце медианой по отрицательным значениям. Затем приведем столбец к абсолютным значениям.

In [16]:
data.loc[(data['days_employed']>=0) | (data['days_employed'].isna() == True),'days_employed']= negative_median
data['days_employed']=abs(data['days_employed'])
display(data.head())

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0    1630.019381         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [17]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Все пропуски и не вызывающие доверия значения заменены на медианные значения трудового стажа трудоустроенных клиентов. Все значения приведены в абсолютные.

## Шаг 2.3. Изменение типов данных.

В столбце `total_income`, отображающем ежемесечный доход клиентов, данные представлены вещественнымтипом данных. Для облегчения работы с ними преобразуем их в целочисленные значения.

In [18]:
data['total_income'] = data['total_income'].astype(int)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int32  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(5)
memory usage: 1.9+ MB


## Шаг 2.4. Удаление дубликатов.

В данных могут присутсвовать дубликаты.Проведем поэтапный анализ и избавимся от них.

### Явные дубликаты

Определим количество явных дубликатов в датафрейме.

In [19]:
data.duplicated().sum()

54

Избавимся от явных дубликатов методом `drop_duplicates()`

In [20]:
data = data.drop_duplicates().reset_index(drop=True)
data.duplicated().sum()

0

Все явные дубликаты удалены из датафрейма.

### Неявные дубликаты

Поищем неявные дубликаты в каждом из необработанных столбцов

**Столбец `education`**

In [21]:
data ['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

В данном столбце обнаружены дубликаты, отличающиеся друг от друга регистром. Все дубликаты можно изменить, приведя столбец к нижнему регистру.

In [22]:
data ['education'] = data ['education'].str.lower()
data ['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

**Столбец `children`**

In [23]:
data ['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

В столбце, отображающем количество детей у клиента, вызывают сомннение значения `-1` и `20`. Можно предположить что `-` и `0` были введены случайно при заполнении. Заменим значения на `1` и `2` соответсвенно методом `replace`.

In [24]:
wrong_numbers = [-1, 20]
right_numbers = [1, 2]
data ['children'] = data ['children'].replace(wrong_numbers, right_numbers)

In [25]:
data ['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

**Столбец `dob_years`**

In [26]:
data ['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75], dtype=int64)

Исходя из данных, представляющих количество клиентов банка каждого возраста, можно с уверенностью сказать, что в столбце `dob_years` присутствуют значение с возрастом `0`, что является очевидной ошибкой в данных. Заменим возраст `0` на ближайший минимальный возраст клиента, определив его.

In [27]:
data.sort_values('dob_years').groupby('dob_years')['dob_years'].count().head(2)

dob_years
0     101
19     14
Name: dob_years, dtype: int64

Ближайший подходящий минимальный возраст - 19 лет заменим методом `replace`.

In [28]:
data ['dob_years'] = data ['dob_years'].replace(0, 19)
data.sort_values('dob_years').groupby('dob_years')['dob_years'].count().head()

dob_years
19    115
20     51
21    111
22    183
23    253
Name: dob_years, dtype: int64

**Столбец `education_id`**

In [29]:
data ['education_id'].unique()

array([0, 1, 2, 3, 4], dtype=int64)

Все хорошо

**Столбец `family_status`**

In [30]:
data ['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Дубликатов нет, но есть значение, выделяющееся по регистру. Приведем столбец к общему нижнему регистру.

In [31]:
data ['family_status'] = data ['family_status'].str.lower()
data ['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'не женат / не замужем'], dtype=object)

**Столбец `family_status_id`**

In [32]:
data ['family_status_id'].unique()

array([0, 1, 2, 3, 4], dtype=int64)

Все хорошо

**Столбец `gender`**

In [33]:
data ['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

В столбце `gender` есть одно выбивающееся из общей картины значение `XNA`. Так как данный столбец не планируется использовать в дальнейшем анализе, и нет возможности с уверенностью отнести указанного клиента к какому-либо из гендеров, примем решение оставить данное значение без изменений.

**Столбец `income_type`**

In [34]:
data ['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

Все хорошо

**Столбец `debt`**

In [35]:
data ['debt'].unique()

array([0, 1], dtype=int64)

Все хорошо

**Столбец `purpose`**

In [36]:
data ['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

В столбце `purpose` введено большое количество уникальных значений, схожих по смыслу. Вместо приведения их к ограниченному количеству уникальных значений, в дальнейшем произведем их категоризацию.

### Методы обработки дубликатов

Для обнаружения **явных** дубликатов использовался метод `.duplicated()`, для их удаления использовался метод `.drop_duplicates()`.

Подобные дубликаты могут возникнуть в процессе копирования, либо выгрузки данных, что может говорить как о человечском факторе, так и о технологической ошибке.

Для обнаружения **неявных** дубликатов использовался метод `.unique()`.
В данном датафрейме неявные дубликаты были представлены значениями, различающиеся регистром и некорректным написанием. 
- Для приведения к общему регистру использовался метод `.str.lower()`. 
- ДЛя устранения ошибок в введенных данных использовался метод `.replace()`.

Неявные дубликаты чаще всего возникают из-за человеческого фактора.

## Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Для удобства работы с датафреймом можно произвести декомпозицию исходного датафрейма при помощи создания словарей. Тем более что в исходном датафрейме присутсвуют столбцы `education_id` и `family_status_id`, позволяющие создать словари на их основе.

Создадим два новых датафрейма, скопировав в них столбцы из исходного, избавившись от дубликатов методом `.drop_duplicates()` и обновив индексацию методом `.reset_index(drop=True)`.

In [37]:
education_dict=data[['education','education_id']].drop_duplicates().reset_index(drop=True).sort_values('education_id')
family_status_dict =data[['family_status','family_status_id']].drop_duplicates().reset_index(drop=True).sort_values('family_status_id')

In [38]:
display(education_dict)
display(family_status_dict)

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

Удалим из исходного датафрейма столбцы `education` и `family_status`.

In [39]:
data.drop(columns=['education','family_status'],axis=1,inplace=True)

In [40]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21471 entries, 0 to 21470
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21471 non-null  int64  
 1   days_employed     21471 non-null  float64
 2   dob_years         21471 non-null  int64  
 3   education_id      21471 non-null  int64  
 4   family_status_id  21471 non-null  int64  
 5   gender            21471 non-null  object 
 6   income_type       21471 non-null  object 
 7   debt              21471 non-null  int64  
 8   total_income      21471 non-null  int32  
 9   purpose           21471 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(3)
memory usage: 1.6+ MB


## Шаг 2.6. Категоризация дохода.

Для удобства работы с данными о доходах клиентов, может разбить доход на несколько категорий, чтобы в дальнейшем при анализе обращаться к ним.
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.

Для этого напишем функцию `categorise_income`, которая пройдет по значениям столбца `total_income` и в зависимости от значения присвоит столбцу `'total_income_category` значение категории дохода.

In [41]:
def categorise_income(column):
    '''
    0–30000 — \'E\';
30001–50000 — \'D\';
50001–200000 — \'C\';
200001–1000000 — \'B\';
1000001 и выше — \'A\'.
'''
    if column<=30000:
        return 'E'
    if column<=50000:
        return 'D'
    if column<=200000:
        return 'C'
    if column<=1000000:
        return 'B'
    return 'A'

Применим функцию

In [42]:
data['total_income_category']=data['total_income'].apply(categorise_income)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21471 entries, 0 to 21470
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   children               21471 non-null  int64  
 1   days_employed          21471 non-null  float64
 2   dob_years              21471 non-null  int64  
 3   education_id           21471 non-null  int64  
 4   family_status_id       21471 non-null  int64  
 5   gender                 21471 non-null  object 
 6   income_type            21471 non-null  object 
 7   debt                   21471 non-null  int64  
 8   total_income           21471 non-null  int32  
 9   purpose                21471 non-null  object 
 10  total_income_category  21471 non-null  object 
dtypes: float64(1), int32(1), int64(5), object(4)
memory usage: 1.7+ MB


## Шаг 2.7. Категоризация целей кредита.

Как было отмечено ранее в столбце `purpose` введено большое количество уникальных значений, схожих по смыслу. Произведем их категоризацию, создав логическую функцию `purpose_categorisation`, которая пройдет по подстрокам и присвоит столбцу `purpose_category` общие категории целей кредита.

In [43]:
def purpose_categorisation(column):
    try:
        if 'автомобил' in column:
            return 'операции с автомобилем' 
        if 'жил' in column or 'недвиж' in column:
            return 'операции с недвижимостью'
        if 'свад' in column:
            return 'проведение свадьбы'
        if 'образов' in column:
            return 'получение образования'
    except:
        return 'категория не определена'

Применим функцию

In [44]:
data['purpose_category']=data['purpose'].apply(purpose_categorisation)
data['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

## Ответы на вопросы.

### Вопрос 1:

**Есть ли зависимость между количеством детей и возвратом кредита в срок?**

Создадим датафрейм, содержащий только значения из столбцов `children` и `debt`

In [45]:
data_children=data[['children','debt']]
display(data_children.head())

children  debt
0         1     0
1         1     0
2         0     0
3         3     0
4         0     0

Создадим словарь, содержащий, количество клиентов категоризированных по количеству детей.

In [46]:
children_dict = data_children.groupby('children').count()
children_dict = children_dict.rename(columns={'debt': 'creditors_total'})
display(children_dict)

creditors_total
children                 
0                   14107
1                    4856
2                    2128
3                     330
4                      41
5                       9

Соеденим датафрейм со словарем и преобразуем значения нового столбца в доли по категории.

In [47]:
data_children = data_children.merge(children_dict, on='children', how='left')
data_children['creditors_total'] = 1/data_children['creditors_total']
display(data_children.head())

children  debt  creditors_total
0         1     0         0.000206
1         1     0         0.000206
2         0     0         0.000071
3         3     0         0.003030
4         0     0         0.000071

Выведем сводную таблицу, отображающую доли по столбцу `debt` по категориям.

In [48]:
data_pivot_children = data_children.pivot_table(index='children', columns='debt', values=('creditors_total'), aggfunc='sum')
data_pivot_children =data_pivot_children.fillna(0)
display(data_pivot_children.sort_values(by = 0, ascending = False))

debt             0         1
children                    
5         1.000000  0.000000
0         0.924647  0.075353
3         0.918182  0.081818
1         0.908361  0.091639
2         0.905075  0.094925
4         0.902439  0.097561

И выведем сводную таблицу, отображающую количество клиентов по столбцу `debt` по категориям.

In [49]:
data_pivot_children_2 = data.pivot_table(index='children', columns='debt', values='dob_years', aggfunc='count')
display(data_pivot_children_2)

debt            0       1
children                 
0         13044.0  1063.0
1          4411.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

#### Вывод 1:

Зависимость между количеством детей и возвратом кредита в срок в данном датафрейме выражена не очень явно. В связи с тем что клиентов, имеющих больше 2 детей мало, данные по этим клиентам не показательны. Общий процент возврата кредитов по всем категориям колеблется между 92 - 90%. У клиентов с 5 детьми процент возврата кредита 100%, это связано с тем, что таких клиентов всего 9.

### Вопрос 2:

**Есть ли зависимость между семейным положением и возвратом кредита в срок?**

Создадим датафрейм, содержащий только значения из столбцов `family_status_id` и `debt`

In [50]:
data_family=data[['family_status_id','debt']]

Добавим значения статуса семейного положения из ранее созданного словаря `family_status_dict`.

In [51]:
data_family = data_family.merge(family_status_dict, on='family_status_id', how='left')

Создадим словарь, содержащий, количество клиентов категоризированных по семейному положению.

In [52]:
client_number_dict = data_family.groupby('family_status_id').count()
client_number_dict.drop(columns=['debt'],axis=1,inplace=True)
client_number_dict = client_number_dict.rename(columns={'family_status': 'creditors_total'})

Соеденим датафрейм со словарем и преобразуем значения нового столбца в доли по категории.

In [53]:
data_family = data_family.merge(client_number_dict, on='family_status_id', how='left')
data_family['creditors_total'] = 1/data_family['creditors_total']

Выведем сводную таблицу, отображающую доли по столбцу `debt` по категориям.

In [54]:
data_pivot_family = data_family.pivot_table(index='family_status', columns='debt', values='creditors_total', aggfunc='sum')
display(data_pivot_family.sort_values(by = 0, ascending = False))

debt                          0         1
family_status                            
вдовец / вдова         0.934307  0.065693
в разводе              0.928870  0.071130
женат / замужем        0.924579  0.075421
гражданский брак       0.906798  0.093202
не женат / не замужем  0.902491  0.097509

#### Вывод 2:

Зависимость между семейным положением и возвратом кредита в срок в данном датафрейме выражена не очень явно, однако четко просматривается закономерность: наиболее платежеспособными клиентами являются вдовцы и вдовы, следом за ними располагаются разведенные клиенты. Можно сделать вывод, что зависимость между семейным положением и возвратом кредита в срок существует. Зависимость слабая, но в сумме с другими факторами может повлиять на общий процент возврата.

### Вопрос 3:

**Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

Создадим датафрейм, содержащий только значения из столбцов `total_income_category` и `debt`

In [55]:
data_income=data[['total_income_category','debt']]

Создадим словарь, содержащий, количество клиентов категоризированных по уровню дохода.

In [56]:
income_dict = data_income.groupby('total_income_category').count()
income_dict = income_dict.rename(columns={'debt': 'creditors_total'})

Соеденим датафрейм со словарем и преобразуем значения нового столбца в доли по категории.

In [57]:
data_income = data_income.merge(income_dict, on='total_income_category', how='left')
data_income['creditors_total'] = 1/data_income['creditors_total']

Выведем сводную таблицу, отображающую доли по столбцу `debt` по категориям.

In [58]:
data_pivot_income = data_income.pivot_table(index='total_income_category', columns='debt', values='creditors_total', aggfunc='sum')
display(data_pivot_income.sort_values(by = 0, ascending = False))

debt                          0         1
total_income_category                    
D                      0.940000  0.060000
B                      0.929379  0.070621
A                      0.920000  0.080000
C                      0.915175  0.084825
E                      0.909091  0.090909

#### Вывод 3:

Прослеживается закономерность между категории дохода клиента и возратом кредита в срок. 

Категория `D` с доходом `30001–50000` оказалась наиболее ответсвенной при возврате кредита. Это может быть связано с тем что данная категория граждан более рационально подходит к распределению своих средств и имеет достаточно стабильный заработок, чтобы погашать платежи по кредитам ежемесечно в срок и без задержек.  При этом на столь высокий процент возврата кредитов влияет и то, что в случае просрочки платежа данные клиенты рискуют получить штраф, выбивающийся из их месячного бюджета и доставляющий им серьезные неудобства.

Категория `В` с доходом `200001–1000000`, стоящая на 2 месте , скорее всего, представляет собой прослойку менеджмента высшего звена, имеющих внушительный ежемесячный доход, однако заботящихся о своей кредитной истории и репутации, что в свою очередь влияет на высокий процент возврата кредитов.

Категория `А` с доходом `1000001 и выше` скорее всего представлена бизнесом, доход которого напрямую зависит от рынка, в следствие чего нестабилен, чтов влияет на более низкий процент возврата кредитов.

### Вопрос 4:

**Как разные цели кредита влияют на его возврат в срок?**

Создадим датафрейм, содержащий только значения из столбцов `'purpose_category` и `debt`

In [59]:
data_purpose=data[['purpose_category','debt']]

Создадим словарь, содержащий, количество клиентов категоризированных по цели, на которую взят кредит.

In [60]:
purpose_dict = data_purpose.groupby('purpose_category').count()
purpose_dict = purpose_dict.rename(columns={'debt': 'creditors_total'})

Соеденим датафрейм со словарем и преобразуем значения нового столбца в доли по категории.

In [61]:
data_purpose = data_purpose.merge(purpose_dict, on='purpose_category', how='left')
data_purpose['creditors_total'] = 1/data_purpose['creditors_total']


Выведем сводную таблицу, отображающую доли по столбцу `debt` по категориям.

In [62]:
data_pivot_purpose = data_purpose.pivot_table(index='purpose_category', columns='debt', values='creditors_total', aggfunc='sum')
display(data_pivot_purpose.sort_values(by = 0, ascending = False))

debt                             0         1
purpose_category                            
операции с недвижимостью  0.927686  0.072314
проведение свадьбы        0.920343  0.079657
получение образования     0.907823  0.092177
операции с автомобилем    0.906453  0.093547

#### Вывод 4:

Прослеживается четкая закономерность кредиты на `операции с недвижимостью` и `проведение свадеб` возвращают более регулярно. В то время как кредиты на `получение образования` и `операции с автомобилем` возвращают менее охотно.

## Общий вывод:

Существует явная закономерность между процентом возврата кредитов в зависомости от семейного положения, уровня дохода и цели кредита. Если провести анализ более подробно, рассматривая клиентов по нескольким категориям, можно проследить еще больше закономерностей. 

Зависимость прецента возврата кредитов от количество детей клиента прслеживается нечетко. Практически все категории возвращают кредит с сравнительно одинаковой частотой.

Для построения более точной модели *кредитного скоринга* рекомендуем провести более детальный анализ по нескольким категориям, кроме того предоставить болььший объем данных для уточнения анализа закономерности процента возврата кредита от количества детей клиента.